drivers/can/can_loopback.c

```diff
@@ -0,0 +1,225 @@
+			filter->std_id : filter->ext_id;
+	mask = filter->id_type == CAN_STANDARD_IDENTIFIER ?
+			filter->std_id_mask : filter->ext_id_mask;
+
+	return ((id & mask) == (frame_id & mask));
+}
+
+int can_loopback_send(struct device *dev, const struct zcan_frame *frame,
+		      s32_t timeout, can_tx_callback_t callback,
+		      void *callback_arg)
+{
+	struct can_loopback_data *data = DEV_DATA(dev);
+	int i;
+	struct can_loopback_filter *filter;
+
+	LOG_DBG("Sending %d bytes on %s. "
+		    "Id: 0x%x, "
+		    "ID type: %s, "
+		    "Remote Frame: %s"
+		    , frame->dlc, dev->config->name
+		    , frame->id_type == CAN_STANDARD_IDENTIFIER ?
+				frame->std_id : frame->ext_id
+		    , frame->id_type == CAN_STANDARD_IDENTIFIER ?
+		    "standard" : "extended"
+		    , frame->rtr == CAN_DATAFRAME ? "no" : "yes");
+
+	if (!data->loopback) {
+
+		return 0;
+	}
+
+	k_mutex_lock(&data->mtx, timeout);
+	for (i = 0; i < CONFIG_CAN_MAX_FILTER; i++) {
+		filter = &data->filters[i];
+		if (filter->rx_cb) {
+			if (check_filter_match(frame, &filter->filter)) {
+				dispatch_frame(frame, filter);
+			}
+		}
+	}
+
+	k_mutex_unlock(&data->mtx);
+
+	if (callback) {
+		callback(CAN_TX_OK, callback_arg);
+	}
+
+	return 0;
+}
+
+
+static inline int get_free_filter(struct can_loopback_filter *filters)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_CAN_MAX_FILTER; i++) {
+		if (filters[i].rx_cb == NULL) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+int can_loopback_attach_isr(struct device *dev, can_rx_callback_t isr,
+			    void *cb_arg,
+			    const struct zcan_filter *filter)
+{
+	struct can_loopback_data *data = DEV_DATA(dev);
+	struct can_loopback_filter *loopback_filter;
+	int filter_id;
+
+	LOG_DBG("Setting filter ID: 0x%x, mask: 0x%x", filter->ext_id,
+		    filter->ext_id_mask);
+	LOG_DBG("Filter type: %s ID %s mask",
+		    filter->id_type == CAN_STANDARD_IDENTIFIER ?
+			"standard" : "extended",
+		     ((filter->id_type && (filter->std_id_mask == CAN_STD_ID_MASK)) ||
+		     (!filter->id_type && (filter->ext_id_mask == CAN_EXT_ID_MASK))) ?
+			"with" : "without");
+
+	k_mutex_lock(&data->mtx, K_FOREVER);
+	filter_id = get_free_filter(data->filters);
+
+	if (filter_id < 0) {
+		LOG_ERR("No free filter left");
+		k_mutex_unlock(&data->mtx);
+		return CAN_NO_FREE_FILTER;
+	}
+
+	loopback_filter = &data->filters[filter_id];
+
+	loopback_filter->rx_cb = isr;
+	loopback_filter->cb_arg = cb_arg;
+	loopback_filter->filter = *filter;
+	k_mutex_unlock(&data->mtx);
+
+	LOG_DBG("Filter attached. ID: %d", filter_id);
+
+	return filter_id;
+}
+
+void can_loopback_detach(struct device *dev, int filter_id)
+{
+	struct can_loopback_data *data = DEV_DATA(dev);
+
+	LOG_DBG("Detach filter ID: %d", filter_id);
+	k_mutex_lock(&data->mtx, K_FOREVER);
+	data->filters[filter_id].rx_cb = NULL;
+	k_mutex_unlock(&data->mtx);
+}
+
+int can_loopback_configure(struct device *dev, enum can_mode mode,
+				u32_t bitrate)
+{
+	struct can_loopback_data *data = DEV_DATA(dev);
+
+	ARG_UNUSED(bitrate);
```

Comment:
I usually put the unused variables at the first line of the function implementation. First the unused and after that, the code. I don't know if it is a good practice or just my of doing that.

Reply:
I thought the definitions have to be at the beginning.